### PR TITLE
Refine image pipeline: limit to top genes, generalize plot scale (SCP-5399)

### PR DESCRIPTION
**Pull request by eweitz**
This fine-tunes the image pipeline and related frontend code.  It improves reliability, relevance, and tailoring.

Previously (#1683), image pipeline attempted to cache images for _all_ genes.  This is expensive and error-prone.  To improve the reliability of pipeline runs while ensuring that the most important genes in a study show fast expression plots, image pipeline now only generates images for the top 100 genes per https://github.com/broadinstitute/scp-ingest-pipeline/pull/334.  

Ranges for expression, and x and y coordinates, are now tailored to the specific gene.  Before, hardcoded values were used.  Miscellaneous updates, like changing to use a Viridis color scale, are also now in place.

To test:
* Complete steps in https://github.com/broadinstitute/scp-ingest-pipeline/pull/334
* In line 236 of `image-pipeline/expression-scatter-plots.js`, set `const useDataCache = false`
* `cd image-pipeline`
* Run `touch log.txt`
* `yarn install`
* Run:
> IS_LOCAL=1 NODE_TLS_REJECT_UNAUTHORIZED=0 node expression-scatter-plots.js --accession="SCP367" --environment="staging" --bucket="fc-1e02d505-cdc3-4ac2-aa7f-13e2a6d79db6" --cluster="KaZhouAll_mt10_SCP_UMAP_coords_meta.csv" --cores="2"; tput bel
* Confirm success message at end of run

This satisfies SCP-5399.

### CODE REVIEW

**Review by bistline — APPROVED** [review]
Looks good - two small gotchas that happened during testing but nonetheless was able to manually confirm this works:

```
Completed Image Pipeline run.  Status: success.  Total run time: 178969 ms (00:02:58.9)
```
Problems:

1. `Error: ENOENT: no such file or directory, open 'log.txt'` - ran `touch log.txt` to get around this
2. `Failed to fetch:
https://singlecell-staging.broadinstitute.org/single_cell/api/v1/studies/SCP367/explore?isImagePipeline=true` - make sure you're on the VPN and there's no deployment happening!

File: image-pipeline/expression-scatter-plots.js [hunk 1]
```diff
@@ -183,17 +223,38 @@ async function prefetchExpressionData(gene, context) {
...
+    expressionArrayString = await downloadFromBucket(fromFilePath, context)
+
+    if (values['json-dir']) {
+      print('Populated initExpressionResponse for development run', context)
+      return
+    }
+  } else {
+    // Enable bypassing JSON data cache, e.g. for development or special production runs
+    const apiStem = `${fetchOrigin}/single_cell/api/v1`
+    const allFields = 'expression'
```
[the remark this answers is not in the document]
Comment (eweitz) [reply image-pipeline/expression-scatter-plots.js]: I'd trivially adapted this block from a block above -- I've inlined the variable now.

**Comment by eweitz** [comment]
> Error: ENOENT: no such file or directory, open 'log.txt' - ran touch log.txt to get around this

I haven't been able to reproduce this, whether on or off VPN.  I'll keep an eye on this when we run it in other environments.

> Failed to fetch: https://singlecell-staging.broadinstitute.org/single_cell/api/v1/studies/SCP367/explore?isImagePipeline=true - make sure you're on the VPN and there's no deployment happening!

Yup, this was a crude error message.  I've refined it with a tip to connect to VPN in this case!

**Comment by eweitz** [comment]
The [build failure](https://github.com/broadinstitute/single_cell_portal_core/actions/runs/7388830768/job/20100439992?pr=1960#step:5:4351) is a false positive, [also seen in independent changes under review](https://github.com/broadinstitute/single_cell_portal_core/pull/1961#issuecomment-1874393350).

```
Total elapsed time: 29 minutes, 53 seconds

### FAILURES and ERRORS ###
*** Yarn test failures ***

*** Rails test failures ***

  HcaAzulClientTest#test_should_get_HCA_metadata_tsv_link [/home/app/webapp/test/integration/external/hca_azul_client_test.rb:170]:
  Expected: 302
Exiting with code: 1
```

**Comment by eweitz** [comment]
> an ENOENT error about log.txt

Ah, Jon encountered that too.  As noted above, I haven't been able to reproduce this, whether on or off VPN. I'll keep an eye on this when we run it in other environments.  For now I've updated the test instructions with a workaround.

> Here's the fail to fetch and my job status was failure

Your screenshot shows you were running with `environment="development"`.  If you run

> IS_LOCAL=1 NODE_TLS_REJECT_UNAUTHORIZED=0 node expression-scatter-plots.js --accession="SCP367" --environment="staging" --bucket="fc-1e02d505-cdc3-4ac2-aa7f-13e2a6d79db6" --cluster="KaZhouAll_mt10_SCP_UMAP_coords_meta.csv" --cores="2"; tput bel

with the updated test instructions, it should work!

**Review by jlchang — APPROVED** [review]
I was able to run expression-scatter-plots.js successfully against the staging study (with Node 18):

```
Completed Image Pipeline run.  Status: success.  Total run time: 201338 ms (00:03:21.3)


Browser 0: File "output/SCP367/images/GFAP.webp" uploaded to destination "cache/expression_scatter/images/KaZhouAll_mt10_SCP_UMAP_coords_meta_csv/GFAP.webp" in bucket "fc-1e02d505-cdc3-4ac2-aa7f-13e2a6d79db6"
File "log.txt" uploaded to destination "parse_logs/expression_scatter_images__2024-01-08T163448.txt" in bucket "fc-1e02d505-cdc3-4ac2-aa7f-13e2a6d79db6"
```

> Your screenshot shows you were running with `--environment=development`. If you run
> > IS_LOCAL=1 NODE_TLS_REJECT_UNAUTHORIZED=0 node expression-scatter-plots.js --accession="SCP367" --environment="staging" --bucket="fc-1e02d505-cdc3-4ac2-aa7f-13e2a6d79db6" --cluster="KaZhouAll_mt10_SCP_UMAP_coords_meta.csv" --cores="2"; tput bel
> 
> with the updated test instructions, it should work!

I was trying to run against my local instance so the command I used was tailored to my copy local version of production SCP2247. So this was what i was trying:
`IS_LOCAL=1 NODE_TLS_REJECT_UNAUTHORIZED=0 node expression-scatter-plots.js --accession="SCP62" --environment="development" --bucket="fc-3d0921be-fe65-48cd-b678-5eb4747d7253" --cluster="KaZhouAll_mt10_SCP_UMAP_coords_meta.csv" --cores="2"; tput bel`

Running against my local instance of the study (using Node 18) I still get fetch error I screenshotted previously. Should I be able to run against a study in my local instance that I own?

Approving since I'm able to run expression-scatter-plots.js successfully.

**Comment by eweitz** [comment]
> Should I be able to run against a study in my local instance that I own?

Yes, but the commands etc. are a bit different than what's outlined here.  If you'd like to pair to do that, let me know.

**Pull request merged**